File: 343.integer-break.go
/*
 * @lc app=leetcode id=343 lang=golang
 *
 * [343] Integer Break
 */

// @lc code=start
func integerBreak(n int) int {
	dp := make([]int, n+1)
	dp[0], dp[1] = 1, 1

	for i := 0; i <= n; i++ {
		for j := 0; j < i; j++ {
			dp[i] = max(dp[i], j*max(i-j, dp[i-j]))
		}
	}

	return dp[n]
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

// @lc code=end

File: 343.integer-break.cpp
/*
 * @lc app=leetcode id=343 lang=cpp
 *
 * [343] Integer Break
 */

// @lc code=start
class Solution {
 public:
  int integerBreak(int n) {
    if (n == 2) return 1;
    if (n == 3) return 2;
    if (n == 4) return 4;

    int ans = 1;
    while (n > 4) {
      ans *= 3;
      n -= 3;
    }

    ans *= n;

    return ans;
  }
};
// @lc code=end

